File: code/futureforge/group-35/linkedlist.js
//Section worked on by Irek Jenjoe Ben 20/CSC/066

// Linked Lists
// A linked list is a linear data structure consisting of nodes where each node contains a value and a 
// reference (pointer) to the next node in the sequence. Unlike arrays, linked lists do not have indices 
// and elements are not stored in contiguous memory locations.


class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    // Insert a new node at the beginning of the linked list
    insertAtHead(value) {
        const newNode = new Node(value);
        newNode.next = this.head;
        this.head = newNode;
    }

    // Insert a new node at the end of the linked list
    insertAtTail(value) {
        const newNode = new Node(value);
        if (!this.head) {
            this.head = newNode;
            return;
        }
        let current = this.head;
        while (current.next) {
            current = current.next;
        }
        current.next = newNode;
    }

    // Remove the first occurrence of a value from the linked list
    remove(value) {
        if (!this.head) {
            return;
        }
        if (this.head.value === value) {
            this.head = this.head.next;
            return;
        }
        let current = this.head;
        let prev = null;
        while (current) {
            if (current.value === value) {
                prev.next = current.next;
                return;
            }
            prev = current;
            current = current.next;
        }
    }

    // Print the linked list
    print() {
        let current = this.head;
        const values = [];
        while (current) {
            values.push(current.value);
            current = current.next;
        }
        console.log(values.join(' -> '));
    }
}

// Example usage
const linkedList = new LinkedList();
linkedList.insertAtHead(10);
linkedList.insertAtHead(20);
linkedList.insertAtTail(30);

linkedList.print(); // Output: 20 -> 10 -> 30
linkedList.remove(10);
linkedList.print(); // Output: 20 -> 30

File: code/futureforge/group-34/linkedlist.py
#Section worked on by Enah Murphy 20/CSC/174 (Group Leader)

# LinkedList 
# A linked list is a linear data structure where elements are stored in nodes. 
# Each node contains data and a reference (or pointer) to the next node in the sequence, 
# forming a chain-like structure. Linked lists allow dynamic memory allocation and efficient insertion/deletion of elements.

class Node:
    def __init__(self, data):
        """Initialize a node with data and next pointer."""
        self.data = data
        self.next = None  # Initialize next pointer to None


class LinkedList:
    def __init__(self):
        """Initialize an empty linked list with a head node."""
        self.head = None  # Initialize head node to None

    def append(self, data):
        """Append a new node with data to the end of the linked list."""
        new_node = Node(data)  # Create a new node
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def display(self):
        """Display all nodes in the linked list."""
        current = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")


# Example usage of the LinkedList class
llist = LinkedList()  # Create an empty linked list
llist.append(10)  # Append node with data 10
llist.append(20)  # Append node with data 20
llist.append(30)  # Append node with data 30

print("Linked List:")
llist.display()  # Display the linked list